aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,342 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+      bool has_val = ((val & desiredMask) == (desired & desiredMask));
+      Bitwise digit = at::cuda::Bitfield<Bitwise>::getBitfield(val, current_bit, RADIX_BITS);
+      if (has_val) {
+        atomicAdd(&temp_storage.digit_counters[digit], 1);
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // load digit counter to register, one digit per thread
+  static_assert(RADIX_DIGITS <= BLOCK_THREADS, "this kernel requires RADIX_DIGITS <= BLOCK_THREADS");
+  IndexType digit_count = 0;
+  if (tidx < RADIX_DIGITS) {
+    digit_count = temp_storage.digit_counters[tidx];
+  }
+
+  // if blocks_per_slice == 1, there is no need to do cross-block reduction
+  // in this case counts saved at registers instead of global memory
+  if (blocks_per_slice > 1) {
+
+    if (tidx < RADIX_DIGITS) {
+      counts[block_idx * RADIX_DIGITS + tidx] = digit_count;
+    }
+    __threadfence(); // make sure writes are globally visible
+    __syncthreads(); // make sure all writes are finished before update semaphores
+  }
+
+  // the last block of each slice accumulates counters from multiple blocks and updates desired and ks_to_find
+  __shared__ bool s_is_last_block_done;
+
+  if (tidx == 0) {
+    if (blocks_per_slice == 1) {
+      s_is_last_block_done = true;
+    } else {
+      int blocks_finished_old = atomicAdd(&semaphores[slice_idx], 1);
+      s_is_last_block_done = (blocks_finished_old == blocks_per_slice - 1);
+    }
+  }
+
+  __syncthreads();
+
+  if (!s_is_last_block_done)
+    return;
+
+  // accumulates counters from multiple blocks
+  if (tidx < RADIX_DIGITS && blocks_per_slice > 1) {
+    digit_count = 0;
+    for (int blk = 0; blk < blocks_per_slice; ++blk) {
+      digit_count += counts[(slice_idx * blocks_per_slice + blk) * RADIX_DIGITS + tidx];
+    }
+  }
+
+  // compute the block-wide inclusive prefix sum
+  IndexType& digit_count_cumsum = digit_count;
+  BlockPrefixCallbackOp prefix_op(0);
+  BlockScan(temp_storage.scan_storage).InclusiveSum(digit_count, digit_count_cumsum, prefix_op);
+  __syncthreads();
+  // every thread also need the perfix_sum of it's left value for comparison, so save a copy in shared mem
+  if (tidx < RADIX_DIGITS) {
+    temp_storage.digit_count_cumsum[tidx] = digit_count_cumsum;
+  }
+  __syncthreads();
+
+  if (tidx < RADIX_DIGITS) {
+    IndexType digit_count_cumsum_left = (tidx == 0) ? 0 : temp_storage.digit_count_cumsum[tidx - 1];
+
+    // if not the last pass: update desired and ks_to_find
+    // if last pass: write out the kth value
+    if (digit_count_cumsum_left < k_to_find && k_to_find <= digit_count_cumsum) {
+      desired = at::cuda::Bitfield<Bitwise>::setBitfield(desired, tidx, current_bit, RADIX_BITS);
+      if (current_bit > 0) {
+        desires[slice_idx] = desired;
+        ks_to_find[slice_idx] = k_to_find - digit_count_cumsum_left;
+      } else {
+        kthValues[slice_idx] = TopKTypeConfig<T>::deconvert(desired);
+      }
+    }
+  }
+
+  // reset semaphores for the next pass
+  if (tidx == 0) {
```

Comment:
Should this be the following?
```suggestion
  if (tidx == 0 && blk_idx_in_slice == 0) {
```

Reply:
all the blocks except the last block already returned at https://github.com/pytorch/pytorch/pull/71081/files#diff-28b350b5a7085a67679fe1932e1d60a924a1724fef619ed980d0a512af3d8189R332-R333
```
  if (!s_is_last_block_done)
    return;
```

so I think there is no need to use the first block to reset semaphores.